File: javascript/015/015.js
"use strict";

const Big = require("../lib/big.js/big.js");

const binomialCoefficient = function(top, bottom) {
  let numerator = Big(1);

  for (let i = Big(top).minus(bottom).plus(1); i.lte(top); i = i.plus(1)) {
    numerator = numerator.times(i);
  }

  let denominator = Big(1);

  for (let i = Big(2); i.lte(bottom); i = i.plus(1)) {
    denominator = denominator.times(i);
  }

  return numerator.div(denominator);
};

const latticePaths = function(m, n) {
  m = Big(m);

  n = Big(n);

  if (m.lte(n)) {
    return binomialCoefficient(m.plus(n), n);
  } else {
    return binomialCoefficient(m.plus(n), m);
  }
};

console.log(latticePaths(20, 20).toString());

File: c++/015/015.cpp
#include <gmpxx.h>
#include <iostream>

mpz_class binomialCoefficient(mpz_class top, mpz_class bottom) {
  mpz_class i = top - bottom + 1;

  mpz_class numerator = mpz_class(1);

  while (i <= top) {
    numerator *= i;

    i += 1;
  }

  i = mpz_class(2);

  mpz_class denominator = mpz_class(1);

  while (i <= bottom) {
    denominator *= i;

    i += 1;
  }

  return numerator / denominator;
}

mpz_class latticePaths(mpz_class m, mpz_class n) {
  if (m < n) {
    return binomialCoefficient(m + n, n);
  } else {
    return binomialCoefficient(m + n, m);
  }
}

int main() {
  mpz_class m = mpz_class(20);

  mpz_class n = mpz_class(20);

  mpz_class result = latticePaths(m, n);

  std::cout << result << std::endl;

  return 0;
}
